File: data_structures/tries/solutions.py
"""Tries — prefix tree insert, search, and root replacement."""

from typing import Dict, List, Optional


class TrieNode:
    def __init__(self) -> None:
        self.children: Dict[str, TrieNode] = {}
        self.is_word = False


class Trie:
    """Prefix tree. Insert/search/starts_with are O(length of the query)."""

    def __init__(self) -> None:
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for ch in word:
            if ch not in node.children:
                node.children[ch] = TrieNode()
            node = node.children[ch]
        node.is_word = True

    def search(self, word: str) -> bool:
        node = self._walk(word)
        return bool(node and node.is_word)

    def starts_with(self, prefix: str) -> bool:
        return self._walk(prefix) is not None

    def _walk(self, s: str) -> Optional[TrieNode]:
        node = self.root
        for ch in s:
            if ch not in node.children:
                return None
            node = node.children[ch]
        return node


def replace_words(dictionary: List[str], sentence: str) -> str:
    """Replace each word with the shortest dictionary root. O(total chars)."""
    trie = Trie()
    for word in dictionary:
        trie.insert(word)

    def shortest_root(word: str) -> str:
        node = trie.root
        prefix: List[str] = []
        for ch in word:
            if ch not in node.children:
                return word
            node = node.children[ch]
            prefix.append(ch)
            if node.is_word:
                return "".join(prefix)
        return word

    return " ".join(shortest_root(word) for word in sentence.split())

File: data_structures/tries/Solutions.java
import java.util.*;

/** Tries — prefix tree insert, search, and root replacement. */
public class Solutions {

    static class TrieNode {
        final Map<Character, TrieNode> children = new HashMap<>();
        boolean isWord;
    }

    /** Prefix tree. Insert/search/startsWith are O(length of the query). */
    public static class Trie {
        private final TrieNode root = new TrieNode();

        public void insert(String word) {
            TrieNode node = root;
            for (char ch : word.toCharArray()) {
                node = node.children.computeIfAbsent(ch, k -> new TrieNode());
            }
            node.isWord = true;
        }

        public boolean search(String word) {
            TrieNode node = walk(word);
            return node != null && node.isWord;
        }

        public boolean startsWith(String prefix) {
            return walk(prefix) != null;
        }

        String shortestRoot(String word) {
            TrieNode node = root;
            StringBuilder prefix = new StringBuilder();
            for (char ch : word.toCharArray()) {
                node = node.children.get(ch);
                if (node == null) return word;
                prefix.append(ch);
                if (node.isWord) return prefix.toString();
            }
            return word;
        }

        private TrieNode walk(String s) {
            TrieNode node = root;
            for (char ch : s.toCharArray()) {
                node = node.children.get(ch);
                if (node == null) return null;
            }
            return node;
        }
    }

    /** Replace each word with the shortest dictionary root. O(total chars). */
    public static String replaceWords(List<String> dictionary, String sentence) {
        Trie trie = new Trie();
        for (String word : dictionary) trie.insert(word);
        String[] words = sentence.split(" ");
        for (int i = 0; i < words.length; i++) {
            words[i] = trie.shortestRoot(words[i]);
        }
        return String.join(" ", words);
    }
}
